waiter/src/waiter/auth/authentication.clj

```diff
@@ -201,3 +225,126 @@
               handler' (middleware/wrap-merge handler auth-params-map)]
           (handler' request))
         (handler request)))))
+
+(defn process-auth-expires-at-request
+  "Handler to allow a client to update its knowledge of when a user's cookie-based credentials expire.
+   Returns a json response containing the expires-at key containing the expiration time in UTC epoch seconds.
+   Relies on the metadata in the x-waiter-auth cookie."
+  [password {:keys [headers]}]
+  (let [decoded-auth-cookie (get-and-decode-auth-cookie-value headers password)
+        [auth-principal _ auth-metadata] decoded-auth-cookie
+        {:keys [expires-at]} auth-metadata
+        cookie-valid? (decoded-auth-valid? decoded-auth-cookie)
+        sanitized-expires-at (or (when cookie-valid? expires-at) 0)]
+    (log/info "waiter auth cookie parsed"
+              (cond-> {:cookie-valid? cookie-valid? :expires-at expires-at}
+                cookie-valid?
+                (assoc :auth-principal auth-principal)))
+    (utils/attach-waiter-source (utils/clj->json-response {:expires-at sanitized-expires-at
+                                                           :principal auth-principal}))))
+
+(defn attach-authorization-headers
+  "Attaches authentication description headers into the response."
+  [{:keys [authorization/method authorization/principal authorization/user] :as response}]
+  (update response :headers
+          (fn [headers]
+            (cond-> headers
+              method (assoc "x-waiter-auth-method" (name method))
+              principal (assoc "x-waiter-auth-principal" (str principal))
+              user (assoc "x-waiter-auth-user" (str user))))))
+
+(let [auth-keep-alive-done-parameter "done=true"
+      auth-keep-alive-done-parameter-replacement "waiter-redirect=true"]
+
+  (defn remove-done-param-from-keep-alive-https-redirect
+    "Removes the done=true query parameter from the response when it is an https redirect to the keep-alive endpoint.
+     Keeping the done=true parameter would prevent authentication from being triggered for soon-to-expire cookies after
+     the requests comes back with https.
+     To simplify processing of the query string, we replace the done=true parameter with another parameter,
+     waiter-redirect=true, reflecting that the request was redirected."
+    [{:keys [headers status] :as response}]
+    (let [{:strs [location]} headers
+          location-uri (when-not (str/blank? location)
```

Comment:
Upstream handler adds the `location` header.  Sitting downstream, we're parsing that header to determine whether or not upstream actually is performing a redirect instead of performing auth.  It would be better to do either:

1. Move https redirection downstream and out of the auth layer which is currently upstream, avoiding https concerns here entirely, since the redirect will have already handled and this code can assume HTTPS, or
2. More explicitly indicate that https redirection is occurring with a response map entry like `:waiter/redirecting-to-https` or `:waiter/authentication-performed`, adding the done parameter only when `:waiter/authentication-performed` is present.

My preference is for the former.

Reply:
Will move https redirects to a higher level filter, e.g. the current https redirect handler in a separate PR.